### PR TITLE
Convert all main binaries to pflag library

### PR DESCRIPTION
**Pull request by jbeda**
This changes all of our main binaries (except for kubecfg) to the pflag library.  This paves the way toward creating a combine binary.

### CODE REVIEW

**Comment by jbeda** [comment]
@lavalamp -- you want to take a look here?

**Comment by jbeda** [comment]
Note that I'm 99% sure that I'll probably break someone someplace.  Please think of other places where we may have a dependency on this.

**Comment by lavalamp** [comment]
Can you run e2e?

**Comment by jbeda** [comment]
Already run -- e2e passes.

**Comment by jbeda** [comment]
I'll fix the gendocs travis break and push another commit.

File: cmd/kube-proxy/proxy.go [hunk 1]
```diff
@@ -52,6 +53,7 @@ func init() {
 }
 
 func main() {
+	util.AddAllFlagsToPFlags()
```
Comment (thockin) [inline cmd/kube-proxy/proxy.go]: should we capture this and flag.Parse into util.InitFlags() to go with InitLogs() ?

Comment (lavalamp) [inline cmd/kube-proxy/proxy.go]: If you change it to be called just plain Init, that sounds OK.

File: pkg/version/verflag/verflag.go [hunk 2]
```diff
@@ -66,6 +66,10 @@ func (v *versionValue) String() string {
 	return fmt.Sprintf("%v", bool(*v == VersionTrue))
 }
 
+func (v *versionValue) Type() string {
```
Comment (lavalamp) [inline pkg/version/verflag/verflag.go]: Godoc comment

**Comment by lavalamp** [comment]
Actually it looks pretty good. Is someone set up to run e2e on vagrant, too?

**Comment by jbeda** [comment]
I'll run e2e on vagrant before this goes in.

File: cmd/kube-proxy/proxy.go (hunk 1 above)
Comment (jbeda) [inline cmd/kube-proxy/proxy.go]: Will do.

File: pkg/version/verflag/verflag.go (hunk 2 above)
Comment (jbeda) [inline pkg/version/verflag/verflag.go]: Heh - none of the other functions in this file have them :)

File: cmd/kube-proxy/proxy.go (hunk 1 above)
Comment (jbeda) [inline cmd/kube-proxy/proxy.go]: Add added a new file and `util.InitFlags`.   There are places where we do some other stuff between initing flags and logs and so I'm going to leave it for now.

**Comment by jbeda** [comment]
*sigh*  I've been trying to get vagrant e2e to run all afternoon.  I'm left with `goe2e.sh` and `services.sh` failing.  I'm 90% sure it isn't me but I'll continue to debug tomorrow.

**Comment by derekwaynecarr** [comment]
I would like us to get vagrant e2e functional in some form.  It was busted earlier this week.  Ping me on IRC tomorrow and I am happy to help work through this with you.

**Comment by lavalamp** [comment]
I'd vote for rebasing this and getting it in, and then figuring out vagrant e2e separately, since it seems like it's not working for other reasons.

**Comment by jbeda** [comment]
Yeah -- I'll try one more time to run vagrant e2e and then get this in.  I'll address vagrant e2e things separately.

**Comment by filbranden** [comment]
This broke `--version`, it used to be somewhat custom because I recognized both `--version` and `--version=raw` but that doesn't seem to be possible anymore with pflag... I'll probably rework that to accept `--version` and `--raw-version` as boolean flags now, I'll send a PR to follow up on this.

I also tried the new kubelet on containervm and it didn't start it on boot because /etc/default/kubelet contains arguments with a single dash, so we'll need to make sure that that gets fixed on the next containervm release as well @dchen1107 who might be interested.

I can see the motivation for this change but I'm not sure it was really worth it... Especially given that the pflag module itself seems to have some shortcomings, for instance I find it annoying that long flags always require "=" separator for the value and that there's no included support for repeated flags... All in all, I find the state of flag parsing in Go quite depressing, especially considering prior art seems to have been largely ignored. :-(

I'll see if I can manage to address some of the shortcomings of pflag in some PRs to that project.

Cheers,
Filipe

**Comment by jbeda** [comment]
@filbranden Darn -- I knew it broke `--version` and I was going to look for that and note it.  See https://github.com/ogier/pflag/issues/11

pflag isn't perfect, but it (a) is used by the commander framework we are using and it is (b) much more familiar to users.  I'm assuming we can send PRs that way.

The main motivation here is to get to the point where we can start collapsing all of our binaries.  Having two flag libraries was going to make that more complex.

**Comment by filbranden** [comment]
Yes I agree on the upsides.

I can see two possible improvements and I'll look into possibly writing some patches for it...
1. Flags with value should *also* recognize a value in a separate argument instead of same argument with an "=" separator. For instance, both `--flag=value` and `--flag value` should be accepted. GNU getopt works that way.
2. I wish there was an official way to handle repeated flags, e.g. `--address=127.0.0.1 --address=10.244.0.1 --address=1.2.3.4` and return a `[]string` from them. I think that would be a nice addition to pflag and would potentially allow us to get rid of `util.StringList`, assuming we're willing to get rid of `,` as a value separator (which IMO is a bad idea, repeated args are certainly preferrable.)

As I said, I'll try to address that in `pflag` and send PRs upstream, let's see if we can make progress with that...

I'm looking at `--version` and `--raw-version` now, will send you a PR for it shortly.

Cheers,
Filipe

**Pull request merged**